File: Python/celulares_builder.py
class Celular:
    def __init__(self):
        self.items = []
        
    def add(self, item):
        self.items.append(item)
    
    def get_total_cost(self):
        return sum(item.get_cost() for item in self.items)
    
    def __str__(self):
        text = "Celular:\n" 
        text += "\n".join(item.name for item in self.items)
        text += f"\nTotal cost: {self.get_total_cost():.2f}"
        return text       
        
class Item:
    def __init__(self, name, cost):
        self.name = name
        self.cost = cost
                
    def get_cost(self):
        return self.cost  
        
    def __str__(self):
        return f"{self.name}: {self.cost}"

class CelularBuilder:
    def __init__(self):
        self.meal = Celular()
    
    def add_main_item(self):
        pass
    
    def add_gama_item(self):
        pass
    
    def add_modelo_item(self):
        pass
    
    def add_color_item(self):
        pass
    
    def reset(self):
        self.meal = Celular()
    
    def get_meal(self):
        return self.meal

class StandarCelularBuilder(CelularBuilder):
    def add_main_item(self):
        self.meal.add(Item("Marca Iphone", 600))
        
    def add_gama_item(self):
        self.meal.add(Item(" Gama Alta", 300))
        
    def add_modelo_item(self):        
        self.meal.add(Item("generacion 14", 200))

    def add_color_item(self):
        self.meal.add(Item("Color blanco", 300))

class CalidadPrecioBuilder(CelularBuilder):
    def add_main_item(self):
        self.meal.add(Item("Marca Alcatel", 68.95))
        
    def add_gama_item(self):
        self.meal.add(Item(" Gama Media", 25.76))
        
    def add_modelo_item(self):        
        self.meal.add(Item("Modelo Pixy 4", 19.99))
    
    def add_color_item(self):
        self.meal.add(Item("Color Rojo Nacarado", 2.99))

class CelularDirector:
    def __init__(self, builder):
        self.builder = builder
        
    def construct_celular(self):
        self.builder.reset()
        self.builder.add_main_item()
        self.builder.add_gama_item()
        self.builder.add_modelo_item()
        self.builder.add_color_item()

# **Listas para almacenar los celulares construidos**
gama_alta_celulares = []
calidad_precio_celulares = []

# **Construcción de celulares de gama alta**
standard_builder = StandarCelularBuilder()
standard_director = CelularDirector(standard_builder)

for i in range(3):  # Crear 3 celulares de gama alta
    standard_director.construct_celular()
    gama_alta_celulares.append(standard_builder.get_meal())  # Cambiado a get_meal()

# **Construcción de celulares de calidad precio**
calidad_precio_builder = CalidadPrecioBuilder()
calidad_precio_director = CelularDirector(calidad_precio_builder)

for i in range(3):  # Crear 3 celulares de calidad precio
    calidad_precio_director.construct_celular()
    calidad_precio_celulares.append(calidad_precio_builder.get_meal())  # Cambiado a get_meal()

# **Imprimir la lista de celulares de gama alta**
print("\n=== Celulares de Gama Alta ===")
for celular in gama_alta_celulares:
    print(celular)
    print("-" * 30)

# **Imprimir la lista de celulares de calidad precio**
print("\n=== Celulares de Calidad Precio ===")
for celular in calidad_precio_celulares:
    print(celular)
    print("-" * 30)

File: Java/src/main/java/com/celulares_builder/celulares_builder.java
package com.celulares_builder;

import java.util.ArrayList;
import java.util.List;

// Clase Item que representa un artículo (como el modelo de celular, color, etc.)
class Item {
    private String name;
    private double cost;

    public Item(String name, double cost) {
        this.name = name;
        this.cost = cost;
    }

    public double getCost() {
        return cost;
    }

    @Override
    public String toString() {
        return name + ": " + cost;
    }
}

// Clase Celular que representa el objeto que se está construyendo
class Celular {
    private List<Item> items = new ArrayList<>();

    public void add(Item item) {
        items.add(item);
    }

    public double getTotalCost() {
        double total = 0;
        for (Item item : items) {
            total += item.getCost();
        }
        return total;
    }

    @Override
    public String toString() {
        StringBuilder text = new StringBuilder("Celular:\n");
        for (Item item : items) {
            text.append(item).append("\n");
        }
        text.append("Total cost: ").append(String.format("%.2f", getTotalCost()));
        return text.toString();
    }
}

// Clase abstracta CelularBuilder para construir diferentes tipos de celulares
abstract class CelularBuilder {
    protected Celular celular = new Celular();

    public abstract void addMainItem();
    public abstract void addGamaItem();
    public abstract void addModeloItem();
    public abstract void addColorItem();

    public Celular getCelular() {
        return celular;
    }

    public void reset() {
        celular = new Celular();
    }
}

// Implementación concreta del builder para un celular de gama alta
class StandardCelularBuilder extends CelularBuilder {
    @Override
    public void addMainItem() {
        celular.add(new Item("Marca Iphone", 600));
    }

    @Override
    public void addGamaItem() {
        celular.add(new Item(" Gama Alta", 300));
    }

    @Override
    public void addModeloItem() {
        celular.add(new Item("Generación 14", 200));
    }

    @Override
    public void addColorItem() {
        celular.add(new Item("Color blanco", 300));
    }
}

// Implementación concreta del builder para un celular de calidad precio
class CalidadPrecioBuilder extends CelularBuilder {
    @Override
    public void addMainItem() {
        celular.add(new Item("Marca Alcatel", 68.95));
    }

    @Override
    public void addGamaItem() {
        celular.add(new Item(" Gama Media", 25.76));
    }

    @Override
    public void addModeloItem() {
        celular.add(new Item("Modelo Pixy 4", 19.99));
    }

    @Override
    public void addColorItem() {
        celular.add(new Item("Color Rojo Nacarado", 2.99));
    }
}

// Clase CelularDirector que utiliza el builder para construir el celular
class CelularDirector {
    private CelularBuilder builder;

    public CelularDirector(CelularBuilder builder) {
        this.builder = builder;
    }

    public void constructCelular() {
        builder.reset();
        builder.addMainItem();
        builder.addGamaItem();
        builder.addModeloItem();
        builder.addColorItem();
    }
}

public class celulares_builder {
    public static void main(String[] args) {
        // Listas para almacenar los celulares construidos
        List<Celular> gamaAltaCelulares = new ArrayList<>();
        List<Celular> calidadPrecioCelulares = new ArrayList<>();

        // Construcción de celulares de gama alta
        StandardCelularBuilder standardBuilder = new StandardCelularBuilder();
        CelularDirector standardDirector = new CelularDirector(standardBuilder);

        for (int i = 0; i < 3; i++) {
            standardDirector.constructCelular();
            gamaAltaCelulares.add(standardBuilder.getCelular());
        }

        // Construcción de celulares de calidad precio
        CalidadPrecioBuilder calidadPrecioBuilder = new CalidadPrecioBuilder();
        CelularDirector calidadPrecioDirector = new CelularDirector(calidadPrecioBuilder);

        for (int i = 0; i < 3; i++) {
            calidadPrecioDirector.constructCelular();
            calidadPrecioCelulares.add(calidadPrecioBuilder.getCelular());
        }

        // Imprimir la lista de celulares de gama alta
        System.out.println("\n=== Celulares de Gama Alta ===");
        for (Celular celular : gamaAltaCelulares) {
            System.out.println(celular);
            System.out.println("-".repeat(30));
        }

        // Imprimir la lista de celulares de calidad precio
        System.out.println("\n=== Celulares de Calidad Precio ===");
        for (Celular celular : calidadPrecioCelulares) {
            System.out.println(celular);
            System.out.println("-".repeat(30));
        }
    }
}